src/register.c

```diff
@@ -1232,14 +1247,27 @@ op_yank(oparg_T *oap, int deleting, int mess)
 		break;
 
 	    case MLINE:
-		if ((y_current->y_array[y_idx] =
-					    vim_strsave(ml_get(lnum))) == NULL)
+		y_current->y_array[y_idx].length = ml_get_len(lnum);
+		if ((y_current->y_array[y_idx].string =
+					vim_strnsave(ml_get(lnum),
+					y_current->y_array[y_idx].length)) == NULL)
+		{
+		    y_current->y_array[y_idx].length = 0;
 		    goto fail;
+		}
 		break;
 
 	    case MCHAR:
 		{
+		    int  tmp;
+
 		    charwise_block_prep(oap->start, oap->end, &bd, lnum, oap->inclusive);
+
+		    // make sure bd.textlen is not longer than the text
+		    tmp = (int)STRLEN(bd.textstart);
+		    if (tmp < bd.textlen)
+			bd.textlen = tmp;
```

Comment:
shouldn't that part belong into `charwise_block_prep()`?

Reply:
That'll change the behavior of `getregionpos()`

Reply:
Putting this into charwise_block_prep() bring us back to the consistency issue as pointed out by @zeertzjq above.

Reply:
then we should back this change out, no? Or does this fix an actual bug?

Reply:
This piece of code makes sure that specifically the string length of the register is correct, while maintaining consistent behaviour in other parts so the code that use charwise_block_prep().